gapic/src/main/com/google/gapid/views/PipelineView.java

```diff
@@ -184,202 +192,293 @@ protected void onUiThreadError(Loadable.Message error) {
+            stageComposite.layout();
+            selectedStage = stage.getDebugName();
+            stripComposite.redraw();
+          });
+
+          stageButton.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_BACKGROUND));
+          stageButton.setFont(theme.bigBoldFont());
+          stageButton.setToolTipText(stage.getStageName() + " (" + stage.getDebugName() + ")");
 
-        for (API.Stage stage : stages) {
-          TabItem item = createStandardTabItem(folder, stage.getDebugName());
+          if (!stage.getEnabled()) {
+            stageButton.setEnabled(false);
+            stageButton.addListener(SWT.Paint, e -> {
+              Rectangle areaSize = stageButton.getBounds();
+              e.gc.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_FOREGROUND));
+              e.gc.setLineWidth((int)(areaSize.height * 0.1));
+              e.gc.drawLine(0, areaSize.height / 2, areaSize.width, areaSize.height / 2);
+              e.gc.drawPolygon(new int[] {0, 0, areaSize.width - 1, 0, areaSize.width - 1, areaSize.height - 1, 0, areaSize.height - 1});
+            });
+          } else {
+            stageButton.addListener(SWT.Paint, e -> {
+              Rectangle areaSize = stageButton.getBounds();
+              e.gc.setLineWidth((int)(areaSize.height * 0.1));
+
+              if (hoveredButton == stageButton) {
+                RGB hoverColor = getDisplay().getSystemColor(SWT.COLOR_BLUE).getRGB();
+                stageButton.setBackground(new Color(stageButton.getDisplay(), hoverColor.red+30, hoverColor.green+30, hoverColor.blue, 63));
+              } else {
+                stageButton.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_BACKGROUND));
+              }
 
-          Composite stageGroup = createComposite(folder, new GridLayout());
+              if (hoveredButton == stageButton || (selectedStage != null && selectedStage.equals(stageButton.getText()))) {
+                e.gc.setForeground(getDisplay().getSystemColor(SWT.COLOR_BLUE));
+              } else {
+                e.gc.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_FOREGROUND));
+              }
+              e.gc.drawPolygon(new int[] {0, 0, areaSize.width - 1, 0, areaSize.width - 1, areaSize.height - 1, 0, areaSize.height - 1});
+            });
+
+            stageButton.addListener(SWT.MouseEnter, e -> {
+              hoveredButton = stageButton;
+              stripComposite.redraw();
+            });
+
+            stageButton.addListener(SWT.MouseExit, e -> {
+              hoveredButton = null;
+              stripComposite.redraw();
+            });
+          }
 
-          FillLayout nameLayout = new FillLayout(SWT.VERTICAL);
-          nameLayout.marginHeight = 5;
-          Composite nameComposite = withLayoutData( createComposite(stageGroup, nameLayout),
-              new GridData(SWT.BEGINNING, SWT.BEGINNING, false, false));
+          if (stageIndex != stages.size()-1) {
```

Comment:
Insert the following selection of the initial button before this line:
```
          if (selectedButton == null) {
            selectedButton = stageButton;
            selectedStage = stage.getDebugName();
            stageStack.topControl = stageMap.get(selectedStage);
          }
```

Reply:
Except now every time I change commands, I go straight back to IA/CS instead of the stage I was at.

That was the whole point of using the StageMap. The buttons are disposed each time you switch commands because you may have different pipelines bound, meaning the stage you were on might not exist anymore. The one constant are the names. By using the stage map, I could detect if a stage at the newly selected command no longer existed (the selectedStage string would no longer be in the map) or not, and act accordingly (go the first stage available, whether it is IA or CS).

Reply:
You can change it back and implement another solution to remember and redraw the previously selected button when you press another button. This is what caused the bug on Windows.

Reply:
Okay. After reviewing the code a bit, the best change is to probably get rid of selectedButton and have stageMap map stage name's to an plain object that contains both the composite and the button. The reason for this, is that I can't check if the selected stage from the last pipeline still exists in the new one until after I create all the buttons and stage composites.